synth_waveform_F32.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * AudioSynthWaveform_F32
+ * 
+ * Created: Patrick Radius, January 2017
+ * Purpose: Generate waveforms at a given frequency and amplitude. Allows for pitch-modulation and portamento.
+ *          
+ * This processes a single stream fo audio data (ie, it is mono)       
+ *          
+ * MIT License.  use at your own risk.
+*/
+#ifndef SYNTHWAVEFORMF32_H
+#define SYNTHWAVEFORMF32_H
+
+#include <arm_math.h>
+#include <AudioStream_F32.h>
+
+class AudioSynthWaveform_F32 : public AudioStream_F32
+{
+  public:
+    enum OscillatorMode {
+        OSCILLATOR_MODE_SINE = 0,
+        OSCILLATOR_MODE_SAW,
+        OSCILLATOR_MODE_SQUARE,
+        OSCILLATOR_MODE_TRIANGLE
+    };
+
+    AudioSynthWaveform_F32(void) : AudioStream_F32(1, inputQueueArray_f32),
+                _PI(2*acos(0.0f)),
+                twoPI(2 * _PI),
+                _OscillatorMode(OSCILLATOR_MODE_SINE),
+                _Frequency(440.0f),
+                _Phase(0.0f),
+                _PhaseIncrement(0.0f),
+                _PitchModAmt(0.0f),
+                _PortamentoIncrement(0.0f),
+                _PortamentoSamples(0),
+                _CurrentPortamentoSample(0),
+                _NotesPlaying(0) {};
+
+    void frequency(float32_t freq) {
+        float32_t nyquist = AUDIO_SAMPLE_RATE_EXACT/2;
+
+        if (freq < 0.0) freq = 0.0;
+        else if (freq > nyquist) freq = nyquist;
+
+        if (_PortamentoSamples > 0 && _NotesPlaying > 0) {
+          _PortamentoIncrement = (freq - _Frequency) / (float32_t)_PortamentoSamples;
```

Comment:
Of the different types of portamento that are commonly done, this appears to be a "fixed time" portamento, ie for any size change in pitch, it will take the same amount of time make the transition.  This is as opposed to "fixed rate" or "exponential" (in pitch, not frequency.  The latter is my favorite because it sounds like the capacitor-based hardware portamento in vintage synths.

OK, that was a digression.  Can you add a comment somewhere around here describing the intended portamento behavior ("fixed time" or whatever you think would be clear) so that others know what's going on more easilyl?